#Script for making boxplots of WGBS methylation at 18 DMRs selected for validation - lethal vs non-lethal vs normal
#Generates Figure S4a

#load libraries
library(GenomicRanges)

#set user defined working directory
BASEDIR<-*user_defined_file_path*
setwd(BASEDIR)

#read in methylation data for plotting, generated by script '02_02_DMRmean_heatmap.R'
load("WGBS/results/methGR_analysisDMRs.RData")
#an_DMRs

#read in DMRs that were assayed in multiplex
selecDMR<-read.table("metadata/SelectedRegions.bed")

#restrict methylation data to just the 18 DMRs used in multiplex
an_selecDMR<-an_DMRs[match(selecDMR$V2,an_DMRs@ranges@start)]

#order amplicons as ordered in all manuscript tables, i.e. by chromosomal position
seqlevels(an_selecDMR)<-seqlevels(an_selecDMR[c(1,12,16,17:22,2:11,13:15,23,24)]
an_selecDMR_ord<-sort(an_selecDMR)

#define samples used in analysis 3
tum <- c("X139C", "X1601C", "X349C", "X379C", "X46C", "X514C", "X564C", "X120C", "X1579C", "X174C", "X202C", "X34C", "X361C", "X506C", "X5237C")

#define lethal and non-lethal samples
lethal<-tum[1:7]
nonlethal<-tum[8:15]

#define normal samples
normals<-c("X448N","X1601N","X564N","X508N")

#define which sample in which group
idx_leth<-vector()
for(i in 1:length(lethal)){
  idx_leth[i]<-grep(lethal[i],colnames(an_selecDMR_ord@elementMetadata))
}

idx_nonleth<-vector()
for(i in 1:length(nonlethal)){
  idx_nonleth[i]<-grep(nonlethal[i],colnames(an_selecDMR_ord@elementMetadata))
}

idx_norm<-vector()
for(i in 1:length(normals)){
  idx_norm[i]<-grep(normals[i],colnames(an_selecDMR_ord@elementMetadata))
}

#extract DMR methylation data at DMRs for plotting 
DMR<-as.matrix(an_selecDMR_ord@elementMetadata)

DMRord<-DMR[,c(idx_leth,idx_nonleth,idx_norm)]

#define number of amplicons to appear on plot
seq_leth2<-seq(from=1,by=7,length.out=18)
seq_nonleth2<-seq(from=2,by=7,length.out=18)
seq_norm2<-seq(from=3,by=7,length.out=18)

#export plot for Figure S4a
pdf("WGBS/plots/MeanWGBSMethBoxplot_bygroup.pdf",width=15,height=5, useDingbats=FALSE)
plot(seq_leth2,DMRord[,1]*100,ylim=c(0,100),xlim=c(0,max(seq_norm2)),xlab="",ylab="DNA methylation (%)",xaxt="n",pch=16, col="white",cex=0.8)

for(i in 1:length(seq_leth2)){
  boxplot(DMRord[i,1:7]*100,boxwex=1.5,at=seq_leth2[i],ylim=c(0,100),xaxt="n",xlab="",ylab="DNA Methylation (%)",add=T,outline=F,yaxt="n",col="dark red")
}
for(i in 1:length(seq_nonleth2)){
  boxplot(DMRord[i,8:15]*100,boxwex=1.5,at=seq_nonleth2[i],ylim=c(0,100),xaxt="n",xlab="",ylab="DNA Methylation (%)",add=T,outline=F,yaxt="n",col="red")
}
for(i in 1:length(seq_norm2)){
  boxplot(DMRord[i,16:19]*100,boxwex=1.5,at=seq_norm2[i],ylim=c(0,100),xaxt="n",xlab="",ylab="DNA Methylation (%)",add=T,outline=F,yaxt="n",col="blue")
}
a<-rep("DMR",18)
b<-1:18
lab<-paste(a,b,sep="")
axis(side=1, at=seq_leth2+1,labels=lab,las=2)
dev.off()
